python/tank/platform/engine.py

```diff
@@ -452,6 +487,95 @@ def destroy_engine(self):
+
+        # Make sure that this engine is configured to run in the new context,
+        # and that it's the EXACT same engine. This can be handled by comparing
+        # the current engine's descriptor to the one coming from the new environment.
+        # If this fails then it's more than just the engine not supporting the
+        # context change, it's that the target context isn't configured properly.
+        # As such, we'll let any exceptions (mostly TankEngineInitError) bubble
+        # up since it's a critical error case.
+        (new_env, engine_descriptor) = _get_env_and_descriptor_for_engine(
+            engine_name=self.instance_name,
+            tk=self.tank,
+            context=new_context,
+        )
+
+        # Make sure that the engine in the target context is the same as the current
+        # engine. If the descriptors are the same object then we know for sure we're
+        # good to go. Otherwise we need to do a little more work before we can feel
+        # safe to continue.
+        if engine_descriptor is not self.descriptor:
+            self.log_debug("Engine %s does not match descriptors between %s and %s." % (
+                self.instance_name,
+                str(self.context),
+                str(new_context)
+            ))
+            if engine_descriptor.get_path() == self.descriptor.get_path():
+                # If the descriptors are different but the install location for
+                # the engine remains the same, then we can feel pretty safe. This
+                # is likely due to the engine running off of a dev descriptor.
+                self.log_debug("Engine descriptor disk paths match, allowing context change.")
+            else:
+                # Since the descriptors differ and the install path for the engine
+                # in the target context doesn't match, we have to disallow the change.
+                raise TankContextChangeNotAllowedError()
+
+        # Run the pre_context_change method to allow for any engine-specific
+        # prep work to happen.
+        self.pre_context_change(self.context, new_context)
+
+        # Check to see if all of our apps are capable of accepting
+        # a context change. If one of them is not, then we remove it
+        # from the persistent app pool, which will force it to be
+        # rebuilt when apps are loaded later on.
+        non_compliant_app_paths = []
+        for install_path, app in self.__application_pool.iteritems():
+            if not app.context_change_allowed:
+                self.log_debug(
+                    "App %s does not allow context changes and will be restarted." % app.name
+                )
+                non_compliant_app_paths.append(install_path)
+            else:
+                # Run the pre-change method on each app to give them
+                # the opportunity to run any app-specific preprocessing.
+                app.pre_context_change(self.context, new_context)
```

Comment:
again maybe more debug logging to wrap these callbacks may be more useful. Maybe you feel this is too excessive. I am thinking of the use case where you have some weird client behaviour that we cannot reproduce - in that case having a big debug dump where we can see exactly when each app ran its init, got torn down etc, in detail can be useful to pinpoint where things are going weird.